File: src/dotnet/ApiSdk/IFlatFileReader.cs
namespace ApiSdk;

/// <summary>
/// Interface for reading flat file JSON data
/// </summary>
public interface IFlatFileReader
{
    /// <summary>
    /// Reads a JSON file from the specified path and returns the content as a string
    /// </summary>
    /// <param name="filePath">The path to the JSON file</param>
    /// <returns>The file content as a string</returns>
    /// <exception cref="ArgumentException">Thrown when the file path is invalid</exception>
    /// <exception cref="FileNotFoundException">Thrown when the file does not exist</exception>
    /// <exception cref="UnauthorizedAccessException">Thrown when access to the file is denied</exception>
    /// <exception cref="IOException">Thrown when an I/O error occurs</exception>
    Task<string> ReadFileAsync(string filePath);

    /// <summary>
    /// Reads a JSON file from the specified path and deserializes it to the specified type
    /// </summary>
    /// <typeparam name="T">The type to deserialize the JSON to</typeparam>
    /// <param name="filePath">The path to the JSON file</param>
    /// <returns>The deserialized object</returns>
    /// <exception cref="ArgumentException">Thrown when the file path is invalid</exception>
    /// <exception cref="FileNotFoundException">Thrown when the file does not exist</exception>
    /// <exception cref="JsonException">Thrown when JSON deserialization fails</exception>
    Task<T> ReadFileAsync<T>(string filePath);

    /// <summary>
    /// Validates that the provided file path is valid
    /// </summary>
    /// <param name="filePath">The file path to validate</param>
    /// <returns>True if the path is valid, false otherwise</returns>
    bool ValidatePath(string filePath);
}

File: src/js/src/interfaces/IFlatFileReader.ts
/**
 * Interface for reading flat file JSON data
 */
export interface IFlatFileReader {
  /**
   * Reads a JSON file from the specified path and returns the content as a string
   * @param filePath The path to the JSON file
   * @returns Promise resolving to the file content as a string
   * @throws {InvalidFilePathError} When the file path is invalid
   * @throws {FileNotFoundError} When the file does not exist
   * @throws {FileReadError} When an I/O error occurs
   */
  readFile(filePath: string): Promise<string>;

  /**
   * Reads a JSON file from the specified path and parses it
   * @param filePath The path to the JSON file
   * @returns Promise resolving to the parsed JSON object
   * @throws {InvalidFilePathError} When the file path is invalid
   * @throws {FileNotFoundError} When the file does not exist
   * @throws {JsonParseError} When JSON parsing fails
   */
  readFileAsJson<T = any>(filePath: string): Promise<T>;

  /**
   * Validates that the provided file path is valid
   * @param filePath The file path to validate
   * @returns True if the path is valid, false otherwise
   */
  validatePath(filePath: string): boolean;
}

